File: app/src/main/java/com/druger/aboutwork/presenters/MyReviewsPresenter.java
package com.druger.aboutwork.presenters;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.druger.aboutwork.db.FirebaseHelper;
import com.druger.aboutwork.interfaces.view.MyReviewsView;
import com.druger.aboutwork.model.Company;
import com.druger.aboutwork.model.Review;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

import static com.druger.aboutwork.db.FirebaseHelper.getCompanies;
import static com.druger.aboutwork.db.FirebaseHelper.getReviews;

/**
 * Created by druger on 09.05.2017.
 */

@InjectViewState
public class MyReviewsPresenter extends MvpPresenter<MyReviewsView> implements ValueEventListener {

    private DatabaseReference dbReference;
    private ValueEventListener valueEventListener;

    private List<Review> reviews = new ArrayList<>();

    public void fetchReviews(String userId) {
        getViewState().showProgress(true);
        reviews.clear();
        dbReference = FirebaseDatabase.getInstance().getReference();

        Query reviewsQuery = INSTANCE.getReviews(dbReference, userId);
        reviewsQuery.addValueEventListener(this);
    }

    @Override
    public void onDataChange(DataSnapshot dataSnapshot) {
        fetchReviews(dataSnapshot);
    }

    @Override
    public void onCancelled(DatabaseError databaseError) {

    }

    private void fetchReviews(DataSnapshot dataSnapshot) {

        for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
            final Review review = snapshot.getValue(Review.class);
            if (!reviews.contains(review)) {
                Query queryCompanies = INSTANCE.getCompanies(dbReference, review.getCompanyId());
                valueEventListener = new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot dataSnapshot) {
                        if (dataSnapshot.exists()) {
                            for (DataSnapshot data : dataSnapshot.getChildren()) {
                                Company company = data.getValue(Company.class);
                                review.setName(company.getName());
                            }
                        }
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {
                        
                    }
                };
                queryCompanies.addValueEventListener(valueEventListener);
                review.setFirebaseKey(snapshot.getKey());
                reviews.add(review);
            }
        }
        getViewState().showProgress(false);
        getViewState().showReviews(reviews);
    }

    public void removeListeners() {
        dbReference.removeEventListener(this);
        if (valueEventListener != null) {
            dbReference.removeEventListener(valueEventListener);
        }
    }

    public Review getReview(int position) {
        return reviews.get(position);
    }

    public void addReview(int position, Review review) {
        addToFirebase(review);
    }

    public void removeReview(int position) {
        FirebaseHelper.INSTANCE.removeReview(reviews.remove(position).getFirebaseKey());
    }

    public void addDeletedReviews(List<Review> deletedReviews) {
        reviews.addAll(deletedReviews);
    }

    public void addToFirebase(Review review) {
        FirebaseHelper.INSTANCE.addReview(review);
    }
}

File: app/src/main/java/com/druger/aboutwork/presenters/MyReviewsPresenter.kt
package com.druger.aboutwork.presenters

import com.druger.aboutwork.App
import com.druger.aboutwork.db.FirebaseHelper
import com.druger.aboutwork.interfaces.view.MyReviewsView
import com.druger.aboutwork.model.Company
import com.druger.aboutwork.model.Review
import com.druger.aboutwork.utils.Analytics
import com.google.firebase.database.*
import moxy.InjectViewState
import moxy.MvpPresenter
import timber.log.Timber
import java.util.*
import javax.inject.Inject

/**
 * Created by druger on 09.05.2017.
 */

@InjectViewState
class MyReviewsPresenter @Inject constructor() : MvpPresenter<MyReviewsView>(), ValueEventListener {

    @Inject
    lateinit var analytics: Analytics

    private lateinit var dbReference: DatabaseReference
    private var valueEventListener: ValueEventListener? = null

    private val reviews = ArrayList<Review>()

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        App.appComponent.inject(this)
    }

    fun fetchReviews(userId: String) {
        viewState.showProgress(true)
        reviews.clear()
        dbReference = FirebaseDatabase.getInstance().reference

        val reviewsQuery = FirebaseHelper.getReviewsById(dbReference, userId)
        reviewsQuery.addValueEventListener(this)
    }

    override fun onDataChange(dataSnapshot: DataSnapshot) {
        fetchReviews(dataSnapshot)
    }

    override fun onCancelled(databaseError: DatabaseError) {
        viewState.showProgress(false)
        Timber.e(databaseError.toException())
    }

    private fun fetchReviews(dataSnapshot: DataSnapshot) {

        if (dataSnapshot.exists()) {
            for (snapshot in dataSnapshot.children) {
                val review = snapshot.getValue(Review::class.java)
                if (!reviews.contains(review)) {
                    review?.companyId?.let { id ->
                        val queryCompanies = FirebaseHelper.getCompanies(dbReference, id)
                        valueEventListener = object : ValueEventListener {
                            override fun onDataChange(dataSnapshot: DataSnapshot) {
                                if (dataSnapshot.exists()) {
                                    for (data in dataSnapshot.children) {
                                        val company = data.getValue(Company::class.java)
                                        review.name = company?.name
                                        viewState.updateAdapter()
                                    }
                                }
                            }

                            override fun onCancelled(databaseError: DatabaseError) {
                                Timber.e(databaseError.message)
                                viewState.showProgress(false)
                            }
                        }
                        queryCompanies.addValueEventListener(valueEventListener as ValueEventListener)
                        review.firebaseKey = snapshot.key
                        reviews.add(review)
                    }
                }
            }
            viewState.showProgress(false)
            viewState.showReviews(reviews)
        } else {
            viewState.showProgress(false)
            viewState.showEmptyReviews()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        removeListeners()
    }

    private fun removeListeners() {
        valueEventListener?.let {
            dbReference.removeEventListener(this)
            dbReference.removeEventListener(it)
        }
    }

    fun getReview(position: Int): Review {
        return reviews[position]
    }

    fun addReview(position: Int, review: Review) {
        addToFirebase(review)
    }

    fun removeReview(position: Int) {
        reviews.removeAt(position).firebaseKey?.let { FirebaseHelper.removeReview(it) }
    }

    fun addDeletedReviews(deletedReviews: List<Review>) {
        reviews.addAll(deletedReviews)
    }

    fun addToFirebase(review: Review) {
        FirebaseHelper.addReview(review)
    }

    fun logEvent(event: String) {
        analytics.logEvent(event)
    }
}
